test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java

```diff
@@ -142,19 +172,172 @@ public void testPstmtInvalidation() throws Throwable
 
         createTable("CREATE TABLE %s (key int primary key, val int)");
 
+        long initialEvicted = numberOfEvictedStatements();
+
         for (int cnt = 1; cnt < 10000; cnt++)
         {
             prepareStatement("INSERT INTO %s (key, val) VALUES (?, ?) USING TIMESTAMP " + cnt, clientState);
 
-            if (numberOfEvictedStatements() > 0)
+            if (numberOfEvictedStatements() - initialEvicted > 0)
             {
+                assertEquals("Number of statements in table and in cache don't match", numberOfStatementsInMemory(), numberOfStatementsOnDisk());
+
+                // prepare more statements to trigger more evictions
+                for (int cnt2 = cnt + 1; cnt2 < cnt + 10; cnt2++)
+                    prepareStatement("INSERT INTO %s (key, val) VALUES (?, ?) USING TIMESTAMP " + cnt2, clientState);
+
+                // each new prepared statement should have caused an eviction
+                assertEquals("eviction count didn't increase by the expected number", 10, numberOfEvictedStatements() - initialEvicted);
+                assertEquals("Number of statements in memory (expected) and table (actual) don't match", numberOfStatementsInMemory(), numberOfStatementsOnDisk());
+
                 return;
             }
         }
 
         fail("Prepared statement eviction does not work");
     }
 
+    /**
+     * Invoked by byteman whenever QueryProcessor.evictPreparedStatement is called in testAsyncPstmtInvalidation
+     */
+    @SuppressWarnings("unused")
+    private static void recordEvictionTimestamp(MD5Digest key)
```

Comment:
Could this be removed and call `put` directly as you do with `preparedStatementLoadTimestamps`?

Reply:
🤔 maybe, i'll see if I can make it work.  The main reason I did this was because we need to use `ClientState.getTimestamp()` to generate a timestamp to record when this was invoked (since we don't use explicit timestamps on the delete)  But perhaps I can make this works with lambdas, or alternative change `evictPreparedStatement` to require a client timestamp.

Reply:
was able to make this work, thanks!